pkg/kube/multicluster/cluster.go

```diff
@@ -46,69 +46,95 @@ type Cluster struct {
 	initialSyncTimeout *atomic.Bool
 }
 
+type ACTION int
+
+const (
+	Add ACTION = iota
+	Update
+)
+
+func (a ACTION) String() string {
+	switch a {
+	case Add:
+		return "Add"
+	case Update:
+		return "Update"
+	}
+	return "Unknown"
+}
+
 // Run starts the cluster's informers and waits for caches to sync. Once caches are synced, we mark the cluster synced.
 // This should be called after each of the handlers have registered informers, and should be run in a goroutine.
-func (r *Cluster) Run(mesh mesh.Watcher, handlers []handler, buildComponents func(cluster *Cluster)) {
+func (c *Cluster) Run(mesh mesh.Watcher, handlers []handler, action ACTION) {
 	if features.RemoteClusterTimeout > 0 {
 		time.AfterFunc(features.RemoteClusterTimeout, func() {
-			if !r.initialSync.Load() {
-				log.Errorf("remote cluster %s failed to sync after %v", r.ID, features.RemoteClusterTimeout)
-				timeouts.With(clusterLabel.Value(string(r.ID))).Increment()
+			if !c.initialSync.Load() {
+				log.Errorf("remote cluster %s failed to sync after %v", c.ID, features.RemoteClusterTimeout)
+				timeouts.With(clusterLabel.Value(string(c.ID))).Increment()
 			}
-			r.initialSyncTimeout.Store(true)
+			c.initialSyncTimeout.Store(true)
 		})
 	}
+
 	// Build a namespace watcher. This must have no filter, since this is our input to the filter itself.
 	// This must be done before we build components, so they can access the filter.
-	namespaces := kclient.New[*corev1.Namespace](r.Client)
-	filter := filter.NewDiscoveryNamespacesFilter(namespaces, mesh, r.stop)
-	kube.SetObjectFilter(r.Client, filter)
-
-	buildComponents(r)
+	namespaces := kclient.New[*corev1.Namespace](c.Client)
+	// This will start a namespace informer and wait for it to be ready. So we must start it in a go routine to avoid blocking.
+	filter := filter.NewDiscoveryNamespacesFilter(namespaces, mesh, c.stop)
+	kube.SetObjectFilter(c.Client, filter)
 
-	if !r.Client.RunAndWait(r.stop) {
-		log.Warnf("remote cluster %s failed to sync", r.ID)
+	for _, h := range handlers {
+		switch action {
+		case Add:
+			h.clusterAdded(c)
+		case Update:
+			h.clusterUpdated(c)
+		}
+	}
+	if !c.Client.RunAndWait(c.stop) {
+		log.Warnf("remote cluster %s failed to sync", c.ID)
 		return
 	}
 	for _, h := range handlers {
-		if !kube.WaitForCacheSync("cluster"+string(r.ID), r.stop, h.HasSynced) {
-			log.Warnf("remote cluster %s failed to sync handler", r.ID)
+		// TODO: we may only need to check this cluster's cache
```

Comment:
not sure what this means, do we not want to wait for each component to be fully synced? I think we do to ensure everything is ready before we mark Istiod ready

Reply:
I mean this is only called on cluster add/update, we only need to check synced for this particular cluster, rather than all the clusters. Not a costly though in common. But if one existing cluster is provided with bad kubeconfig, then it will take more time for a new cluster initiation.

Reply:
https://github.com/istio/istio/pull/49730